### PR TITLE
Codomain maps render url

### PR DESCRIPTION
**Pull request by will-moore**
# What this PR does

Updates the url query string for encoding codomain 'maps'.
See https://trello.com/c/PzMTIpY2/266-rendering-settings-in-url and original PR: https://github.com/openmicroscopy/openmicroscopy/pull/4839

This makes the encoding of 'maps' more extensible for addition of more maps in the future.
We now encode maps as a list of dictionaries. Each dictionary applies to a channel and each dictionary has a key that is an identifier for the map, and the value is another dictionary.
```"reverse"``` codomain only has an "enabled" boolean. For a 2-channel image this looks like:
```
?maps=[{"reverse":{"enabled":true}},{"reverse":{"enabled":false}}
```

# Testing this PR

1. Test rendering of images in Preview and Full viewer. Turn on/off reverse intensity and check that this is rendered correctly.

2. Copy & paste settings with Reverse-Intensity in various combinations of 
     - Copy via 'Copy' button in Preview panel OR via right-click menu on jsTree OR in full viewer
     - Paste via 'Paste' button in Preview panel OR via 'Paste and Save' on jsTree OR in full viewer

# Related reading

Link to cards, tickets, other PRs:

1. https://trello.com/c/PzMTIpY2/266-rendering-settings-in-url

2. Discussion at https://github.com/openmicroscopy/openmicroscopy/pull/4839

cc @chris-allan @waxenegger

### CODE REVIEW

**Comment by will-moore** [comment]
@jburel I'll fix Travis, add tests and re-list... Doesn't need reviewing today (unless feedback on description)

**Comment by waxenegger** [comment]
Tested on http://web-dev-merge.openmicroscopy.org.

In general it worked well, both when applying and copy and paste and in preview as well as full viewer (incl multiple selection and tree).

I have 2 cases, however, when a js error is caused and I suspect it's for the same reason. In both cases the rendering settings are supplied via the url that has all the parameters including map:

1) Hit the button "Full Viewer" in the preview tab then that image is opened with the "full" url.
2) Get the link to the image and open it in a new tab

Seems that the contents of the map param arrive at the following line as an unparsed string but it wants to work with the parsed array content already:

image

**Comment by will-moore** [comment]
@waxenegger Thanks. Fixed and added a robot test for that.

**Comment by waxenegger** [comment]
Yes, that fixed it, opening with a url and the "Full viewer" button works fine now.

codomain_url

Though the robot doesn't seem to like your test case: https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/513/robot/

**Comment by will-moore** [comment]
Last commit should fix [currently failing Test Launch Full Viewer](https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/516/robot/)

**Comment by mtbc** [comment]
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/517/robot/ shows "Test Launch Full Viewer" passing and "Test Rating" newly failing.

**Comment by will-moore** [comment]
Last commit should fix failing robot test of Firefox at https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/520/robot/Web%20&%20Web/Web_1/Rdef%20Test/Test%20LUTs%20Reverse%20Intensity/

**Comment by jburel** [comment]
https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/521/robot/Web%20&%20Web/Web_1/Rdef%20Test/Test%20LUTs%20Reverse%20Intensity/ is now passing

File: components/tests/ui/testcases/web/rdef_test.txt [hunk 1]
```diff
@@ -373,7 +409,101 @@ Test Rdef Save
...
+    Wait For Image Src                      maps=[{%22reverse%22:{%22enabled%22:true}
+    Wait For Image Src                      16_colors
+
+    # Paste and Save by right-click on next Image in Tree
+    Click Next Thumbnail
+    Wait For Image Src                      FF0000
+    ${imageId2}=                            Get Selected Id From Tree
+    ${imgNodeId2}=                          Wait For Image Node         ${imageId2}
+    Right Click Rendering Settings          ${imgNodeId2}               Paste and Save
+    Wait For Image Src                      16_colors
```
Comment (jburel) [inline components/tests/ui/testcases/web/rdef_test.txt]: did you check with invalid lut name?
Do we correctly fall back on color?

**Comment by jburel** [comment]
Have you try passing an invalid values in the url e.g. lut=foo?

**Comment by jburel** [comment]
The change will be "less" disruptive
Tests are passing

**Comment by waxenegger** [comment]
yes, tests are passing (https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/522/console) and the maps parameter is at the end of the requests on http://web-dev-merge.openmicroscopy.org

**Comment by waxenegger** [comment]
I might not be a bad idea to wrap the JSON.parse call in a try catch in case users go "evil" or map contents are accidentally messed up programmatically, just so that the image request remains valid. Of course one could always try things to break the url in other places if one wants to...

image

**Comment by will-moore** [comment]
@waxenegger Yes, I guess JSON.parse is more sensitive to errors than the rest of the render string so I'll add a try/catch.

**Comment by will-moore** [comment]
@waxenegger That last commit should give you an alert with invalid json for ```?maps=```

**Comment by waxenegger** [comment]
I'm thinking 'alert' is perhaps a bit on the nose, I find the console log functions less disruptive to the user while still being helpful to the developers. Of course one could say the user might want to know that there is something wrong with the url so it boils down to premise/preference. 

Other than that everything works fine and the corresponding robot test is passing as well.

**Comment by jburel** [comment]
We can downgrade if people complain and think ``alert`` is too much

Ideally we should reset the url to its "working" state but not a blocker

**Pull request merged**